### PR TITLE
Save/Vault offboarding

### PR DESCRIPTION
**Pull request by chrisjgf**
[sc-9032]

## Changelog
- `unwrapAndSend` on `Unwrapper`. Upgradeable contract to unwrap mAsset -> bAsset/fAsset
- `redeemAndUnwrap` on `SavingsContract`. Redeems mAsset & uses `unwrapAndSend`
- `withdrawAndUnwrap` on `BoostedVault`. Withdraws from vault and uses `redeemAndUnwrap`
- `_reduceRaw` on `BoostedTokenWrapper`. Decrements user's raw balance without transferring token

### CODE REVIEW

File: contracts/interfaces/IBoostedVaultWithLockup.sol [hunk 1]
```diff
@@ -17,6 +17,8 @@ interface IBoostedVaultWithLockup {
      */
     function stake(address _beneficiary, uint256 _amount) external;
 
+    // TODO add referrer overload?
```
Comment (alsco77) [inline contracts/interfaces/IBoostedVaultWithLockup.sol:20]: Probably not necessary for the vault

File: contracts/rewards/boosted-staking/BoostedTokenWrapper.sol [hunk 2]
```diff
@@ -132,10 +132,18 @@ contract BoostedTokenWrapper is InitializableReentrancyGuard {
...
-        _rawBalances[msg.sender] -= _amount;
+        _reduceRaw(_amount);
         stakingToken.safeTransfer(msg.sender, _amount);
     }
 
+    /**
+     * @dev Reduced a given staked balance of sender (no transfer)
+     * @param _amount Units of StakingToken
+     */
+    function _reduceRaw(uint256 _amount) internal nonReentrant {
```
Comment (alsco77) [inline contracts/rewards/boosted-staking/BoostedTokenWrapper.sol:143]: adding this second `nonReentrant` will cause any calls to `_withdrawRaw` to fail

File: contracts/rewards/boosted-staking/BoostedVault.sol [hunk 3]
```diff
@@ -249,6 +250,45 @@ contract BoostedVault is
...
+        address _router,
+        bool _isBassetOut
+    ) external override updateReward(msg.sender) updateBoost(msg.sender) {
+        require(_amount > 0, "Cannot withdraw 0");
+
+        // Reduce raw balance (but do not transfer `stakingToken`)
+        _reduceRaw(_amount);
+
+        // Approve SavingsContract to spend this contract's `stakingToken`
+        IERC20(stakingToken).approve(address(stakingToken), _amount);
```
Comment (alsco77) [inline contracts/rewards/boosted-staking/BoostedVault.sol:277]: Do we need to approve here? I think the savingscontract can just burn the tokens from msg.sender (which will be this staking contract account when called below)

File: contracts/savings/SavingsContract.sol [hunk 4]
```diff
@@ -49,6 +50,9 @@ contract SavingsContract is ISavingsContractV2, Initializable, InitializableToke
     event Poked(uint256 oldBalance, uint256 newBalance, uint256 interestDetected);
     event PokedRaw();
 
+    // Tracking events
+    event Referral(address indexed referrer, address beneficiary);
```
Comment (alsco77) [inline contracts/savings/SavingsContract.sol:54]: Think we need to track the amount deposited here

File: contracts/savings/peripheral/Unwrapper.sol [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+     * @param _router         masset or feederpool
+     * @param _input          input token address
+     * @param _output         output token address
+     * @param _amount         amount
+     * @param _minAmountOut   min amount
+     * @param _beneficiary    beneficiary
+     * @return outputQuantity Units of credits burned from sender
+     */
+    function unwrapAndSend(
+        bool _isBassetOut,
```
Comment (alsco77) [inline contracts/savings/peripheral/Unwrapper.sol:54]: Only thing i see in this fn is that the integrator requires some knowledge about the mStable system to know whether this is true or false. Ideally would just pass the output addr

Comment (alsco77) [reply contracts/savings/peripheral/Unwrapper.sol:54]: Might be hard to do in an immutable and gas efficient way

Comment (alsco77) [reply contracts/savings/peripheral/Unwrapper.sol:54]: Maybe we can just have a view fn here that takes an mAsset addr (or imAsset), an output addr, and then returns a bool for `isBassetOut`. At least integration contracts can look this up off chain before sending the tx if they want

**Review by alsco77 — CHANGES_REQUESTED** [review]
Functionality looks good, we have all the paths covered (vault -> imAsset && imAsset -> b/fAsset) and I think the immutable unwrapper and the arg for `_router` makes sense for future proofing.

Changes needed:
 - fix this broken test suite/CI
 - resolve comments
 - add some simple e2e tests for each of the paths (imAsset -> fAsset, imAsset -> bAsset, vault -> fAsset, vault -> bAsset). You could just make a new test file or use the unwrapper

**Review by alsco77 — APPROVED** [review]

File: test-fork/savings/sc21-upgrade.spec.ts [hunk 6]
```diff
@@ -0,0 +1,492 @@
...
+//  1. Deploy Unwrapper
+//   1.1. Set the Unwrapper address as constant in imUSD Vault
+//  2. Upgrade and check storage
+//   2.1. Vaults
+//   2.2. SavingsContracts
+//  3. Do some unwrapping
+//   3.1. Directly to unwrapper
+//   3.2. Via SavingsContracts
+//   3.3. Via SavingsVaults
+context("Unwrapper and Vault upgrades", () => {
```
Comment (doncesarts) [inline test-fork/savings/sc21-upgrade.spec.ts:79]: @doncesarts  review  probably it should be a unit test

File: contracts/rewards/boosted-staking/BoostedVault.sol [hunk 7]
```diff
@@ -249,6 +250,43 @@ contract BoostedVault is
...
+     * @param _isBassetOut   Route action of redeem/swap
+     */
+    function withdrawAndUnwrap(
+        uint256 _amount,
+        uint256 _minAmountOut,
+        address _output,
+        address _beneficiary,
+        address _router,
+        bool _isBassetOut
+    ) external override updateReward(msg.sender) updateBoost(msg.sender) {
```
[the remark this answers is not in the document]
Comment (naddison36) [reply contracts/rewards/boosted-staking/BoostedVault.sol]: I've gone with `outputQuantity` as the return param

File: contracts/savings/SavingsContract.sol [hunk 8]
```diff
@@ -319,19 +347,75 @@ contract SavingsContract is ISavingsContractV2, Initializable, InitializableToke
...
-    {
+    function redeemAndUnwrap(
+        uint256 _amount,
+        bool _isCreditAmt,
+        uint256 _minAmountOut,
+        address _output,
+        address _beneficiary,
+        address _router,
+        bool _isBassetOut
+    ) external override returns (uint256 creditsBurned, uint256 massetReturned) {
```
[the remark this answers is not in the document]
Comment (naddison36) [reply contracts/savings/SavingsContract.sol]: I've gone with `outputQuantity` as the return param

File: test-fork/savings/unwrapper.spec.ts [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+
+        // approve tokens for router
+        const routers = [alusdFeederPool, hbtcFeederPool]
+        const tokens = [musdAddress, mbtcAddress]
+
+        await unwrapper.connect(governor).approve(routers, tokens)
+    })
+
+    it("Can call getIsBassetOut & it functions correctly", async () => {
+        const isCredit = true
```
[the remark this answers is not in the document]
Comment (naddison36) [reply test-fork/savings/unwrapper.spec.ts:89]: I've added tests for when `isCredit` is false

**Pull request closed without merging**